src/albert/utils/task_utils.py

```diff
@@ -0,0 +1,326 @@
+        )
+
+    block_info = next((blk for blk in (task.blocks or []) if blk.id == block_id), None)
+    if block_info is None:
+        raise ValueError(
+            f"Block '{block_id}' not found on task {task.id} for metadata construction."
+        )
+    data_templates: list[TaskMetadataDataTemplate] = []
+    for dt in block_info.data_template or []:
+        data_templates.append(
+            TaskMetadataDataTemplate(
+                id=getattr(dt, "id", None) or "",
+                name=getattr(dt, "name", None),
+                full_name=getattr(dt, "full_name", None),
+                standards=getattr(dt, "standards", None),
+            )
+        )
+    workflows: list[TaskMetadataWorkflow] = []
+    for wf in block_info.workflow or []:
+        workflows.append(
+            TaskMetadataWorkflow(
+                albert_id=getattr(wf, "id", None),
+                name=getattr(wf, "name", None),
+            )
+        )
+
+    blockdata = TaskMetadataBlockdata(
+        id=block_id,
+        datatemplate=data_templates,
+        workflow=workflows,
+    )
+    return TaskMetadata(
+        filename=filename,
+        task_id=task.id,
+        block_id=block_id,
+        inventories=inventories,
+        blockdata=blockdata,
+    )
+
+
+def resolve_attachment(
+    *,
+    attachment_collection: AttachmentCollection,
+    task_id: str,
+    file_path: str | Path | None,
+    attachment_id: str | None,
+    allowed_extensions: set[str],
+    note_text: str | None,
+) -> str:
+    """Ensure an attachment is available, optionally uploading a new file."""
+
+    if file_path is not None:
+        path = Path(file_path).expanduser()
+        if not path.exists() or not path.is_file():
+            raise FileNotFoundError(f"File not found at '{path}'.")
+        ext = determine_extension(filename=path.name)
+        if allowed_extensions and (ext or "") not in allowed_extensions:
+            raise ValueError(
+                f"File extension {ext} is not permitted. Allowed extensions: {sorted(allowed_extensions)}"
+            )
+        note_text_to_use = note_text or ""
+        with path.open("rb") as file_handle:
+            uploaded_attachment_note = attachment_collection.upload_and_attach_file_as_note(
+                parent_id=task_id,
+                file_data=file_handle,
+                note_text=note_text_to_use,
+                file_name=path.name,
+            )
+        uploaded_attachments = uploaded_attachment_note.attachments or []
+        if not uploaded_attachments:
+            raise ValueError("Failed to upload attachment. Try again.")
+        return uploaded_attachments[0].id
+
+    if attachment_id is None:
+        raise ValueError("attachment_id must be provided when file_path is not supplied.")
+    return attachment_id
+
+
+def fetch_csv_table_rows(
+    *,
+    session: AlbertSession,
+    api_version: str,
+    attachment_id: str,
+    csv_table_key: str | None,
+) -> list[object]:
```

Comment:
Is the type here really not known at all, even a `list[dict]` or `list[str]` or anything?